protocol/messenger_communities.go

```diff
@@ -1947,3 +1953,802 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+					progressUpdates <- importProgress
+					return
+				}
+
+				author.AvatarImagePayload = imagePayload
+				authorProfilesToSave[id] = author
+
+				assetCounter.Increase()
+				progress := float32(assetCounter.Value()) / float32(totalAssetsCount)
+				importProgress.UpdateTaskProgress(discord.DownloadAssetsTask, progress)
+				progressUpdates <- importProgress
+
+				if m.DiscordImportMarkedAsCancelled(discordCommunity.IDString()) {
+					importProgress.StopTask(discord.DownloadAssetsTask)
+					progressUpdates <- importProgress
+					cancel <- discordCommunity.IDString()
+					return
+				}
+			}(id, author)
+		}
+		wg.Wait()
+
+		if m.DiscordImportMarkedAsCancelled(communityID) {
+			importProgress.StopTask(discord.DownloadAssetsTask)
+			progressUpdates <- importProgress
+			cancel <- communityID
+			return
+		}
+
+		for idxRange := range gopart.Partition(len(messageAttachmentsToDownload), 20) {
+			attachments := messageAttachmentsToDownload[idxRange.Low:idxRange.High]
+			wg.Add(1)
+			go func(attachments []*protobuf.DiscordMessageAttachment) {
+				defer wg.Done()
+				for i, attachment := range attachments {
+					assetPayload, contentType, err := discord.DownloadAsset(attachment.Url)
+					if err != nil {
+						errmsg := fmt.Sprintf("Couldn't download message attachment '%s': %s", attachment.Url, err.Error())
+						importProgress.AddTaskError(
+							discord.DownloadAssetsTask,
+							discord.Warning(errmsg),
+						)
+						progressUpdates <- importProgress
+						continue
+					}
+
+					attachment.Payload = assetPayload
+					attachment.ContentType = contentType
+					messageAttachmentsToDownload[i] = attachment
+
+					assetCounter.Increase()
+					// Multiplying progress by `0.8` to leave 20% for saving assets to DB
+					progress := (float32(assetCounter.Value()) / float32(totalAssetsCount)) * 0.8
+					importProgress.UpdateTaskProgress(discord.DownloadAssetsTask, progress)
+					progressUpdates <- importProgress
+
+					if m.DiscordImportMarkedAsCancelled(communityID) {
+						importProgress.StopTask(discord.DownloadAssetsTask)
+						progressUpdates <- importProgress
+						cancel <- communityID
+						return
+					}
+				}
+			}(attachments)
+		}
+		wg.Wait()
+
+		if m.DiscordImportMarkedAsCancelled(communityID) {
+			importProgress.StopTask(discord.DownloadAssetsTask)
+			progressUpdates <- importProgress
+			cancel <- communityID
+			return
+		}
+
+		importProgress.UpdateTaskState(discord.DownloadAssetsTask, discord.TaskStateSaving)
+		progressUpdates <- importProgress
+
+		err = m.persistence.SaveDiscordMessageAttachments(messageAttachmentsToDownload)
+		if err != nil {
+			m.cleanUpImport(communityID)
+			importProgress.AddTaskError(discord.DownloadAssetsTask, discord.Error(err.Error()))
+			importProgress.Stop()
+			progressUpdates <- importProgress
+			return
+		}
+
+		importProgress.UpdateTaskProgress(discord.DownloadAssetsTask, 1)
+		progressUpdates <- importProgress
+
+		err = m.publishOrg(discordCommunity)
```

Comment:
So this publishes the community description and it's also what causes the community to show up Desktop's UI.
This works nice as long as there's mailservers available. In face of a connection error, the community description doesn't reach its own client.

We don't actually need to make the client aware of this community via dispatching messages, but should probably do something similar to MessengerResponse handling.